cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/ConcurrencyConstraint.java

```diff
@@ -17,19 +17,43 @@
 package co.cask.cdap.internal.app.runtime.schedule.constraint;
 
 import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.proto.ProgramRunStatus;
 import co.cask.cdap.proto.ProtoConstraint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * A constraint which dictates an upper bound on the number of concurrent schedule runs.
  */
 public class ConcurrencyConstraint extends ProtoConstraint.ConcurrenyConstraint implements CheckableConstraint {
+  private static final Logger LOG = LoggerFactory.getLogger(ConcurrencyConstraint.class);
 
   public ConcurrencyConstraint(int maxConcurrency) {
     super(maxConcurrency);
   }
 
   @Override
   public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
+    int numRunning = context.getProgramRuns(schedule.getProgramId(), ProgramRunStatus.RUNNING, maxConcurrency).size();
+    if (numRunning >= maxConcurrency) {
+        LOG.debug("Skipping run of program {} from schedule {} because there are at least {} running runs.",
+                  schedule.getProgramId(), schedule.getName(), maxConcurrency);
+      return notSatisfied();
```

Comment:
the `notSatisfied` method below contains that logic.
I just wrapped it into its own method, since its called twice.